From 30c7ea5865a0da859fc03a5ca471a7c7abc5b329 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 28 May 2025 12:24:15 +0200
Subject: [PATCH] firmware: arm_scmi: PM_NOIRQ support over OCALL context

Set OP-TEE SCMI PTA OCALL execution context with non-secure interrupts
masked in OP-TEE between calls to SCMI OP-TEE transport driver PM_NOIRQ
suspend and resume handlers. This permits SCMI OCALL processing during
Linux PM final suspend stages and initial resume stages where Linux
has masked the interrupts at CPU level and may still call SCMI service
as for example to disable/enable clock source, done from Linux syscore
subsystem.

Change-Id: Id4b072caf14696d1efa6f59eed8df37884442077
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/466081
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/firmware/arm_scmi/optee.c | 64 +++++++++++++++++++++++++++++++
 1 file changed, 64 insertions(+)

diff --git a/drivers/firmware/arm_scmi/optee.c b/drivers/firmware/arm_scmi/optee.c
index abca60496e95..05de39689af4 100644
--- a/drivers/firmware/arm_scmi/optee.c
+++ b/drivers/firmware/arm_scmi/optee.c
@@ -212,6 +212,22 @@ enum scmi_optee_ocall_reply {
 	 * and issue Ocall command PTA_SCMI_OCALL_CMD_THREAD_READY again.
 	 */
 	PTA_SCMI_OCALL_PROCESS_MSG = 3,
+	/*
+	 * On return of Ocall, SCMI PTA shall enter PM_NOIRQ state
+	 * that is keeping non-secure interrupts masked for processing
+	 * the next SCMI messages (until PTA_SCMI_OCALL_DISABLE_NOIRQ
+	 * is received) and issuesOcall command PTA_SCMI_OCALL_CMD_THREAD_READY
+	 * again.
+	 */
+	PTA_SCMI_OCALL_ENABLE_NOIRQ = 4,
+	/*
+	 * On return of Ocall, SCMI PTA shall exit PM_NOIRQ state
+	 * that is unmasking back non-secure interrupts masked for processing
+	 * the next SCMI messages (until PTA_SCMI_OCALL_ENABLE_NOIRQ
+	 * is received) and issuesOcall command PTA_SCMI_OCALL_CMD_THREAD_READY
+	 * again.
+	 */
+	PTA_SCMI_OCALL_DISABLE_NOIRQ = 5,
 };
 
 /*
@@ -221,11 +237,13 @@ enum scmi_optee_ocall_reply {
  * @arg: TEE invoke command arguments
  * @param: TEE invoke command parameters
  * @ocall_arg: TEE Ocall2 arguments
+ * @pm_noirq_enabled: True when in PM_NOIRQ state, false otherwise
  */
 struct ocall_ctx {
 	struct tee_ioctl_invoke_arg args;
 	struct tee_param param[4];
 	struct tee_ocall2_arg ocall_arg;
+	bool pm_noirq_enabled;
 };
 
 /**
@@ -1017,6 +1035,51 @@ static int scmi_optee_service_remove(struct device *dev)
 	return 0;
 }
 
+static int scmi_optee_suspend_noirq(struct device *dev)
+{
+	struct scmi_optee_channel *channel;
+
+	list_for_each_entry(channel, &scmi_optee_private->channel_list, link) {
+		struct ocall_ctx *ocall_ctx = channel->ocall_ctx;
+
+		if (!ocall_ctx)
+			continue;
+
+		ocall_ctx->ocall_arg.out_param1 = PTA_SCMI_OCALL_ENABLE_NOIRQ;
+		ocall_ctx->ocall_arg.out_param2 = 0;
+
+		if (invoke_ocall_thread(channel) == 0)
+			ocall_ctx->pm_noirq_enabled = true;
+	}
+
+	return 0;
+}
+
+static int scmi_optee_resume_noirq(struct device *dev)
+{
+	struct scmi_optee_channel *channel;
+
+	list_for_each_entry(channel, &scmi_optee_private->channel_list, link) {
+		struct ocall_ctx *ocall_ctx = channel->ocall_ctx;
+
+		if (!ocall_ctx || !ocall_ctx->pm_noirq_enabled)
+			continue;
+
+		ocall_ctx->ocall_arg.out_param1 = PTA_SCMI_OCALL_DISABLE_NOIRQ;
+		ocall_ctx->ocall_arg.out_param2 = 0;
+
+		if (invoke_ocall_thread(channel) == 0)
+			ocall_ctx->pm_noirq_enabled = true;
+	}
+
+	return 0;
+}
+
+/* Late PM suspend/resume to support handling PM_NOIRQ for SCMI request */
+static const struct dev_pm_ops scmi_optee_pm_ops = {
+	NOIRQ_SYSTEM_SLEEP_PM_OPS(scmi_optee_suspend_noirq, scmi_optee_resume_noirq)
+};
+
 static const struct tee_client_device_id scmi_optee_service_id[] = {
 	{
 		UUID_INIT(0xa8cfe406, 0xd4f5, 0x4a2e,
@@ -1034,6 +1097,7 @@ static struct tee_client_driver scmi_optee_driver = {
 		.bus = &tee_bus_type,
 		.probe = scmi_optee_service_probe,
 		.remove = scmi_optee_service_remove,
+		.pm = pm_ptr(&scmi_optee_pm_ops),
 	},
 };
 
-- 
2.53.0

